src/tangerine/resources/assistant.py

```diff
@@ -235,6 +261,177 @@ def _extract_request_data(self):
+                log.warning(
+                    "AUDIT: Session %s not owned by user %s, starting fresh conversation",
+                    session_uuid,
+                    user_id,
+                )
+                return []
+
+            # Guard against None payload
+            payload = conversation.payload or {}
+            stored_prev_msgs = payload.get("prevMsgs", [])
+            log.info("AUDIT: Retrieved %d messages from stored conversation", len(stored_prev_msgs))
+
+            # Validate stored messages (defensive programming)
+            validated_stored = self._validate_prev_msgs(stored_prev_msgs)
+
+            # Apply rolling window to limit LLM context
+            return self._limit_conversation_to_pairs(validated_stored)
+
+        except Exception as e:
+            log.exception(
+                "Failed to reconstruct conversation history for session %s: %s",
+                session_uuid,
+                str(e),
+            )
+            # Fail gracefully - return empty history rather than breaking the request
+            return []
+
+    def _limit_conversation_to_pairs(self, messages):
+        """
+        Limit conversation history to last 10 human↔assistant Q&A pairs for LLM context efficiency.
+
+        This method:
+        1. Filters for human/assistant messages only (excludes system messages)
+        2. Ensures proper pairing (human question → assistant answer)
+        3. Takes the last 10 complete pairs
+        4. Preserves chronological order
+
+        Args:
+            messages: List of conversation messages
+
+        Returns:
+            Limited list of messages (last 10 Q&A pairs max)
+        """
+        if not messages:
+            return []
+
+        # Filter to only human and assistant messages, preserve order
+        conversation_messages = [msg for msg in messages if msg.get("sender") in {"human", "ai"}]
+
+        if not conversation_messages:
+            return []
+
+        # Extract complete Q&A pairs (human → ai)
+        pairs = []
+        i = 0
+        while i < len(conversation_messages):
+            msg = conversation_messages[i]
+
+            # Look for human message
+            if msg.get("sender") == "human":
+                # Look for corresponding AI response
+                ai_response = None
+                for j in range(i + 1, len(conversation_messages)):
+                    if conversation_messages[j].get("sender") == "ai":
+                        ai_response = conversation_messages[j]
+                        i = j  # Move past this AI response
+                        break
+
+                if ai_response:
+                    pairs.append((msg, ai_response))
+                else:
+                    # Incomplete pair - just the human message at the end
+                    pairs.append((msg,))
+
+            i += 1
+
+        # Take the last 10 pairs maximum
+        if len(pairs) > 10:
```

Comment:
Technically you could get rid of this `if` and just invoke `pairs = pairs[-10:]`

```
>>> my_list = ["a", "b", "c"]
>>> my_list[-10:]
['a', 'b', 'c']
```